### ISSUE TITLE
Limit concurrency of Polly.Cache calling its delegate per corresponding key

### ISSUE THREAD

**Issue by mrmartan** [App-vNext/Polly#657]
**Summary: What are you wanting to achieve?**

I want the cache policy execution to limit concurrency on its data loading delegate per given operation key.

Imagine a disrtibuted cache in Redis implemented though the IAsyncCacheProvider. The cached keys have set expiry/TTL in Redis. The system in question experiences thousands of calls per second. Now the key in Redis expires. There is a high probability that tens of requests come in (nearly) simultaneously all requiring the same key (from Redis). It is not there and the cache executes the underlying data provider once per each request, ie. needlessly because all of those executions will yield the same value. Executing the data provider is expensive. Then all of those executions start setting the same key/value pair to Redis where setting it once would be enough. This IMHO causes unnessesary load on both the cache (Redis) and the backing data store. Or rather bursts of load whenever a key expires.

I would like the provider to be executed only once and once it returns all outstanding/awaiting cache request would be fulfilled immediately, also setting the value into the cache only once.

Do you have any suggestions please? Is this possible with Polly.Cache? Or is this something you would have to implement?

I have tried implementing this using Bulkhead unsuccessfuly. I would have to have two of them - one for the cache provider and one for the data provider - and even then it just limits the conncurency and does not allow for fullfiling all executions in its queue at once.

As I am writing this I thought of using double cache with an inner layer of memory cache but I can't think of how to reasonably tell when it's OK to evict from the inner cache. The problem with Bulkhead is that it does not work on a per key basis and maintaining a collection of them could get expensive quickly. I am also worried about overhead.
Something like Polly.Cache(Redis:IAsyncCacheProvider) -> Bulkhead -> Polly.Cache(Memory:ConcurrentDictionary) -> Bulkhead -> DataLoader

**Comment by reisenberger** [comment]
_Courtesy reply_: I have a collection of ideas how this could be done, but am unlikely to be able to respond in detail before the middle of next week.

**Comment by reisenberger** [comment]
_Some quick input to get you a response_: I would suggest using the `ConcurrentDictionary<TKey, Lazy<TValue>>` pattern ([reference](https://blogs.endjin.com/2015/10/using-lazy-and-concurrentdictionary-to-ensure-a-thread-safe-run-once-lazy-loaded-collection/); [reference](https://andrewlock.net/making-getoradd-on-concurrentdictionary-thread-safe-using-lazy/)) around your DataLoader, to ensure that you only execute the underlying call to Redis once, and can share the result of that call across concurrent calls. `TKey` can be the OperationKey.  

You would want to use the `ConcurrentDictionary` to cache and share the `Lazy` across the burst of concurrent calls, but you would probably want to thread-safely evict that instance of the  `Lazy` from the `ConcurrentDictionary` once it had returned a value (indicating that particular execution cycle on the lazy had finished).

**Comment by phatcher** [comment]
@mrmartan I'd also have a look at https://github.com/aspnet/Extensions/issues/708 which discusses that GetOrAddAsync on IMemoryCache/IDistributedCache it not atomic.

You can also look at my NuGet package ([Meerkat.Caching](https://github.com/phatcher/Meerkat.Caching)) which addresses this and use this as the basis of your own code.

**Comment by reisenberger** [comment]
And there is also [LazyCache](https://github.com/alastairtree/LazyCache) and [DoubleCache](https://github.com/AurumAS/DoubleCache), which it might be possible to [implement a Polly CacheProvider for](https://github.com/App-vNext/Polly/wiki/Implementing-new-cache-providers), or take inspiration from - as well as @phatcher 's example (thanks Paul!).

**Comment by reisenberger** [comment]
@phatcher Aren't most of the `Semaphore`-based approaches in that thread more-or-less just rewriting `Lazy`?

**Comment by phatcher** [comment]
@reisenberger  Possibly, I haven't had a look at the implementation of LazyCache - I came across #708 when I was trying to implement my service (as opposed to HttpRequest) level cache that we discussed a little while ago.

Lazy was a pure in-memory implementation, which is why I took an extension method approach for mine. I still have to see how it behaves in a large scale environment e.g. Kubernetes with pods coming and going.  

What it allows is that the cache key and ISynchronizer key different - which is useful when services could duplicate each other's work, but you can't control the order in which services are invoked.

**Comment by mrmartan** [comment]
Thanks guys. I've went with @phatcher's Synchronizer for now. I like how it builds on top of IMemoryCache. Not the nuget though as it drags dependencies I don't want.

As for `ConcurrentDictionary<TKey, Lazy<TValue>>`: I would have to use something more like `ConcurrentDictionary<TKey, Lazy<Task<TValue>>>`

**Comment by reisenberger** [comment]
@mrmartan Great, glad you have a solution!

Re:

> I would have to use something more like `ConcurrentDictionary<TKey, Lazy<Task<TValue>>>`

Sure, I was making an effort to get you a quick response rather than leave you waiting, I forgot your case was async 🙂 . 

@phatcher nice caching library, thanks for linking.

**Comment by mrmartan** [comment]
I do. I was not using Polly.Cache in the first place and the solution ead me to custom `GetOrSet` implementation. I have a host of other applications that use Polly.Caching though. Do you imagine a way to extend Polly.Caching behavior in a way discussed here? I.e. that the value provider and cache provider set are executed once and atomically.

Once you have time.

**Comment by reisenberger** [comment]
Hi @mrmartan . Yes, it would be with the `ConcurrentDictionary<TKey, Lazy<Task<TValue>>>` approach.

**Comment by reisenberger** [comment]
Linking a related discussion; https://github.com/aspnet/Extensions/issues/708

**Comment by jjxtra** [comment]
Lazy works well if only one box is serving requests, but if you have more than one box, you will still have problems with multiple queries to your database. In this case, locking on a redis key is necessary.

**Comment by jjxtra** [comment]
So I think whatever solution happens here needs to happen at the policy level, not the caching level. A single policy on a single machine with a given key should only ever execute one at a time, any additional call that comes in while that policy is executing should either lock or simply be returned the currently executing `Task` for that policy. So I think this issue is bigger than just caching. If Polly already has a mechanism to do this per key at the policy level, then the caching issue is moot.

**Comment by reisenberger** [comment]
To recap and summarize two possible technical solutions: 

a) **Thread-safe `Lazy`-per-key**: such as the `ConcurrentDictionary<TKey, Lazy<TValue>>` (and equivalent async) pattern [proposed here](https://github.com/App-vNext/Polly/issues/657#issuecomment-504903481). Re (@jjxtra said on #740) "the GetOrCreateAsync pattern perhaps better applies at a higher level then caching" : yes, exactly. I had had the idea that this could become a `ConcurrentDuplicateRequestCollapserPolicy` of general use, regardless of whether (or at what inner level) the governed work includes a `CachePolicy` execution.

b) **Thread-safe lock-per-key**: @phatcher has a great demonstration of this in [Meerkat.Caching](https://github.com/phatcher/Meerkat.Caching) (see foot of readme), as a lock-per-key via `SemaphoreSlim`.  A custom Polly policy or native Polly policy could do similar: perhaps `LockPerKeyPolicy` or `DuplicateRequestLockPolicy`.

In both cases, the key would default to `Context.OperationKey` (or the policy could take an [`ICacheKeyStrategy`](https://github.com/App-vNext/Polly/wiki/Cache#cachekeystrategy-optional)), for similarity with and likely partnering with `CachePolicy`.

Both would be primitives, separate policies which could be composed with cache (or other) policies.

---

A key question for either solution is scoping, ownership and disposal of the semaphores/lazys:

The question-mark over (b) is that the collection of Semaphores can grow over time. Avoiding background jobs to clean this up, the only sensible solution is to hand control of semaphore-eviction-and-disposal back to the user: something like `LockPerKeyPolicy.DisposeKey(string key)`

Option (a) has no such problem: the `Lazy` can be (thread-safely) cleared from the `ConcurrentDictionary` after it has returned a value; all concurrent executions still holding the Lazy will quickly go out of scope, allowing the Lazy to be GCd.  The cost is an allocation of a `Lazy<T>` (32 bytes?) per concurrent invocation set; presumably small in comparison, if the motivation/gain is to avoid the heavy cost of the duplicate guarded operation.  But by not baking this into the CachePolicy by default, users could choose where to apply it.

---

For comparison, [Alastair Crabtree](https://github.com/alastairtree)'s [LazyCache](https://github.com/alastairtree/LazyCache) also addresses the same problem (it's a common caching problem) as a standalone project.  This (afaiu) caches a `Lazy<T>` (or async equivalent). That is suitable for memory caches, but not particularly for serialized-to-json distributed caches, so not generally applicable for Polly.

**Comment by jjxtra** [comment]
I like option `a`, I think having some sort of wrapper policy as you suggest is the way to go, it makes this "opt in" and is easy to apply to any policy with a potentially expensive operation. The added benefit of not maintaining a dictionary of semaphores is also a plus.

A couple of considerations:
- The `GetOrCreateAsync` call still needs to be locked, even with a concurrent dictionary. I've tested this and it's possible for multiple factory methods for the same key to be executed when using concurrent dictionary with `Lazy`. The lock can be a spin lock or other super fast locking mechanism. It just needs to lock long enough for the `Lazy` object to be created and put into the concurrent dictionary. One solution I've done is to make a int array of 8192 ints and use `Interlocked.CompareExchange` on a hash of the key % 8192. If the lock is not acquired, just await a Thread.Yield and try again. Getting a lock should take no more than a microsecond, or less if no `Thread.Yield()` call. The lock can then be released as soon as the `Lazy` has been returned from `GetOrCreateAsync`.
- Having a way to lock on a distributed cache key would be great for making this wrapper policy lock on multiple machines. This is probably one implementation of an interface for the wrapper policy, and another implementation could be a simple in memory lock such as the spin lock I suggest above.
- Have a way to use an in memory cache first before the wrapper policy. A flow like this: `[ In memory cache check ] ->[ Wrapper policy on key [ policy a] -> [ policy b ], etc. ]`. So don't require the wrapper policy to be an all-encompassing policy.

Thanks for being so open and willing to discuss this, I am in charge of a large back-end with some pretty heavy sql queries and am pushing a transition to Polly for our services.

**Comment by jjxtra** [comment]
@reisenberger Have you started any work on this? Any concerns if I take a stab at this? I could create a fork and start doing some work and would love to get some early feedback...

**Comment by reisenberger** [comment]
@jjxtra Awesome if you want to work on this! I already had an early code sketch - will share in the next 24-48 hours

**Comment by jjxtra** [comment]
Looking forward to it, would love collaboration here in order to come up with the best solution.

**Comment by reisenberger** [comment]
Thanks @jjxtra ! Awesome to collaborate.

Just brain-dumped [here](https://github.com/reisenberger/Polly/tree/spike_ConcurrentRequestCollapser) (final commit) the early sketch I had.  Basically a realisation of the pattern [first proposed here](https://github.com/App-vNext/Polly/issues/657#issuecomment-504903481); think this is fairly close to your [first bullet here](https://github.com/App-vNext/Polly/issues/657#issuecomment-570739136) too? (thanks, great discussion!). But please feel free to elaborate / make / propose changes!  :+1: to the idea of interface to inject lock strategies.

That ^ sketch is just an engine for the heart of the sync implementation. The best way to get an idea for the scaffolding for a policy and its syntax is look at how [`NoOpPolicy` is implemented](https://github.com/reisenberger/Polly/tree/spike_ConcurrentRequestCollapser/src/Polly/NoOp) (or others).  We also have a [wiki article on this](https://github.com/App-vNext/Polly/wiki/Polly-concept-and-architecture).

Async version fairly similar with `ConcurrentDictionary<string, Lazy<Task<>>`, do you think?



---

(_Note:_ I am mid working on Polly v8; the sketch branches off that and is best place now to branch from.)

Thanks again!

**Comment by jjxtra** [comment]
Should I fork the repo and then build off of your sketch branch then? Initial ideas for lock providers would be an in-memory per process lock, and a distributed redis lock. The redis lock depends on stack-exchange redis library, so is it better off in the contrib framework instead of the core framework?

**Comment by reisenberger** [comment]
>Should I fork the repo and then build off of your sketch branch then?

Yes, branch from there now.  The on-going [`v800` branch](https://github.com/App-vNext/Polly/tree/v800) may be in motion over coming days, but any re-basing should be limited (/I can pitch in). 

👍 to ideas for lock providers. 

+ Yes, any `ILockProvider` implementation with new dependencies will def become a [Polly-Contrib](https://github.com/Polly-Contrib/) repo. 
+ We have a [blank template repo here](https://github.com/Polly-Contrib/Polly.Contrib.BlankTemplate) for non-policy Polly-Contribs, providing a head start (typical repo content, tfms, ready-made build etc 🙂 )
  - There are a couple of [low-hanging-fruit](https://github.com/Polly-Contrib/Polly.Contrib.BlankTemplate/issues) to bring it up to date with latest Polly 
+ Shout when you have proposed lock providers (or share initially as a GitHub gist or something?- your call), then we can decide for each if it lives in the main Polly or in a Polly-Contrib repo.

Thanks for contributing!

**Comment by jjxtra** [comment]
Here is my idea for the in-process per key lock. It is designed under the assumption that most locks will be acquired and release without contention, which should be the most common path. Acquiring the lock is the cost of computing a hash value from a string, a modulus call and finally an interlocked compare exchange call.

I had considered using `Monitor.Enter` and `Monitor.Exit`, but that had thread implications that concerned me, along with requiring an object per key, which would generate more garbage. I also considered semaphore slim and manual reset event, but again more garbage and worse kernel transitions...

I also considered using `SpinWait` (https://github.com/microsoft/referencesource/blob/master/mscorlib/system/threading/SpinWait.cs) but it seemed to not be net standard 1.1 compliant, and probably overkill for this super tight lock that we want.

Here is the gist, please give me any and all feedback: https://gist.github.com/jjxtra/f6116180b2ef5c1550e60567af506c2a

**Comment by reisenberger** [comment]
Thanks @jjxtra for this and for everything to take `ConcurrentDuplicateRequestCollapser` forward!

>I also made the lock provider interface async, so that any distributed locking mechanism could fit nicely in, but we could change this back if you think we don't need it

Great question.  I think we will need both (separate) `ISyncLockProvider` and `IAsyncLockProvider` interfaces:

+ The synchronous version of the policy will need to consume a synchronous `ISyncLockProvider`.
+ We should provide configuration syntax so that the asynchronous version of the policy can be configured with either an `IAsyncLockProvider` or `ISyncLockProvider` (it may be common to use an in-process lock still on async executions).  When the async policy is configured with a sync `ISyncLockProvider`, we can internally use a lightweight `ValueTask<IDisposable>` wrapper.

With that in mind, I think we can probably move your `ProcessLockProvider` back to a sync implementation.

>I had considered using Monitor.Enter and Monitor.Exit, but that had thread implications that concerned me

With you: `Monitor` cannot be released on a different thread than acquired it.  For [this policy implementation](https://github.com/reisenberger/Polly/commit/d314db4e7b2eb39baac60f43b51134f4e9772e6a), we would only be locking over same-thread manipulation of the in-memory `ConcurrentDictionary<,>`, so `Monitor` would be ok (say if you think I'm missing something). (But this is not to take away from your thread-safe implementation...)

>considered on just locking the string itself, but this would only work if

Agreed :+1:

Still reflecting on where a key-lock implementation like this might eventually live (`Polly`, `Polly.Contrib`?), but hopefully some thoughts ☝️ to take things forward!

**Comment by jjxtra** [comment]
Just a heads up I made some changes to the gist to fix the concurrency issue you mentioned, along with providing a sync and async version. Let me know if you think it's good enough to use or if you have any further suggested changes.

**Comment by jjxtra** [comment]
I think adding the per-process key lock to the main Polly framework would be ideal. It could be the default implementation for the new policy that ensures a key only executes one at a time.

The distributed lock will have dependencies on redis stackexchange, so probably it belongs in the contrib project...

**Comment by reisenberger** [comment]
Thanks @jjxtra , planning to come back to this shortly.

**Comment by reisenberger** [comment]
@jjxtra After [benchmarking and considering complexity trade-offs](https://gist.github.com/jjxtra/f6116180b2ef5c1550e60567af506c2a#gistcomment-3154235), I suggest the default lock implementation should be a single lock per policy, using the standard .Net `lock` / `Monitor` (which the benchmarks shows as suitably performant for the majority use case).  However, a keyed-lock implementation would be a good candidate for Polly-Contrib!

Thanks again for contributing! Let me know if I can provide any other guidance on building out the policy infrastructure ([previous notes here](https://github.com/App-vNext/Polly/issues/657#issuecomment-574071826)).

**Comment by jjxtra** [comment]
@reisenberger I ran some additional tests, was wondering if you wanted to run one last benchmark on your machine using my modified code. I put my findings in the gist link. Interesting results as cpu core count increases.

**Comment by reisenberger** [comment]
@jjxtra I wanted to progress the [main Policy implementation](https://github.com/App-vNext/Polly/issues/657#issuecomment-574396860) for this as well as the locking discussion; I took that up and it is done.  Will push that in the coming days and then revert to this discussion.

**Comment by jjxtra** [comment]
Sounds good!

**Comment by reisenberger** [comment]
In some spare weekend hours I opted to push out the main Policy implementation for this before switching my attention back to Polly v8. The policy implementation is complete (with concurrency-exercising unit tests), added doco just now, [pushed all here](https://github.com/Polly-Contrib/Polly.Contrib.DuplicateRequestCollapser).

The policy has moved to Polly-Contrib at [`Polly.Contrib.DuplicateRequestCollapser`](https://github.com/Polly-Contrib/Polly.Contrib.DuplicateRequestCollapser).  @jjxtra : Just as you had aspects of the striped-locking where supporting .Net Standard 1.x presented challenges, likewise the policy implementation wanted a dependency only available to .Net Standard 2.x.  Moving the policy to Polly-Contrib allows us to make that compromise (we would not drop .Net Standard 1.x support from core Polly just for this policy).

I have [released an early version to nuget](https://www.nuget.org/packages/Polly.Contrib.DuplicateRequestCollapser/), just to get a version of this out there and published.  But @jjxtra : this is just the start, would be great to see this taken further with further locking implementations!  @jjxtra I will send you an invitation to join [Polly-Contrib](https://github.com/Polly-Contrib/), which gives you rights on that repo.  Feel free to expand/change locking implementations if you want to take ownership of that contrib project.  @mrmartan @phatcher If you are interested too in contributing there I can invite you.

Welcome feedback on the early nuget release ( @mrmartan @phatcher @jjxtra / anyone) if you get to exercise it in your environments. Perhaps best over on [`Polly.Contrib.DuplicateRequestCollapser` repo](https://github.com/Polly-Contrib/Polly.Contrib.DuplicateRequestCollapser), given that is where this has ended up!

Thanks!

**Comment by jjxtra** [comment]
I'll do a PR for it. Thanks so much for setting it up and doing all the
documentation, very thorough and impressive work!

-- Jeff

**Comment by reisenberger** [comment]
Np @jjxtra .  

Going to close this issue now that a first version of the policy has been published, but know @jjxtra that you plan to expand on this (thanks!). Thanks to everybody on this thread for the contributions in thinking!

Also linked out from the Polly CachePolicy wiki area to the new policy: https://github.com/App-vNext/Polly/wiki/Avoiding-cache-repopulation-request-storms

**Comment by jjxtra** [comment]
If you have a few seconds to look at my distributed lock implementation, I'd love any feedback. I am using this code on some personal projects, I have no issue licensing the distributed lock bits under your license. It does require stack exchange redis to be brought in, but I think that is a fair trade...

https://github.com/Polly-Contrib/Polly.Contrib.DuplicateRequestCollapser/blob/master/src/Polly.Contrib.DuplicateRequestCollapser/StackexchangeRedisDistributedLockProvider.cs

**Comment by reisenberger** [comment]
Thanks @jjxtra . I'll aim to jump back into this thread in the coming days!

**Comment by reisenberger** [comment]
@jjxtra Good to have a Redis lock in the mix for the request collapser: thanks for bringing this forward!

I'd like to create a new repo for this: `Polly.Contrib.DuplicateRequestCollapser.RedisDistributedLock`.  There may well be users of the `DuplicateRequestCollapserPolicy` who only need the in-memory lock, not Redis; we can avoid those users taking unneeded/unwanted dependencies by hosting the Redis lock in a different repo.  I'll make the new repo, then let's move your Redis code there.

Can I recommend always using [GitFlow](https://guides.github.com/introduction/flow/index.html) on the Polly-Contribs? (ie don't push interim work straight to the `master` branch).  Helps others coming to your Polly-Contrib repo, if branches follow standard expectations.  Pretty sure this is old hat, but in case not, I've [written up in more detail the motivations](https://github.com/Polly-Contrib/Polly.Contrib.CustomPolicyTemplates/blob/master/CONTRIBUTING.md) (if nothing else, at least for others on future projects).

**Comment by reisenberger** [comment]
@jjxtra Repo is ready, with a template: [Polly.Contrib.DuplicateRequestCollapser.RedisDistributedLock](https://github.com/Polly-Contrib/Polly.Contrib.DuplicateRequestCollapser.RedisDistributedLock).  Let's pull your Redis contribution into there!

**Comment by jjxtra** [comment]
Thanks for making the repo, I will do pr moving forward :)

**Comment by coultonluke** [comment]
Is this going to be made available as a nuget package or should we just copy the code into our projects?

**Comment by martincostello** [comment]
That's a question for the repo containing the code - see https://github.com/Polly-Contrib/Polly.Contrib.DuplicateRequestCollapser.RedisDistributedLock/issues/4
